### DISCUSSION TITLE
File-based catalog (FBC) support is live!

### DISCUSSION THREAD

**Discussion by Allda** [redhat-openshift-ecosystem/community-operators-prod#5031] [Announcements]
Hello community!

I am glad to announce that we just released support for a File-based catalog (FBC) for OpenShift operators in this repository.

We have heard your feedback from the last couple of months and realized the current operator workflow has certain limitations and doesn’t allow full operator manipulation.

I am sure not all of you are familiar with what FBC is and its benefits so here are a couple of use cases that you might find useful.

Use cases supported by FBC workflow:

- Direct access to operator update graph
- Bundle channel promotion
- Modification and removal of bundles
- Catalog deprecation feature
- Independent operator bundle and catalog release cycle

More information about FBC can be found in the official upstream documentation.

- https://olm.operatorframework.io/docs/reference/file-based-catalogs/ 
- https://olm.operatorframework.io/docs/reference/catalog-templates/ 

### Where to start?
We prepared documentation with 2 major starting points that cover FBC onboarding. Based on your current status you can follow one of them.

- [I have an existing operator that I would like to convert](https://redhat-openshift-ecosystem.github.io/operator-pipelines/users/fbc_onboarding/)
- [I am starting a new operator and I would like to use all FBC benefits](https://redhat-openshift-ecosystem.github.io/operator-pipelines/users/fbc_workflow/)

We are also going to record a short demo later this week to show you the onboarding process. Subscribe to this thread to get more information and updates.

We are looking for your onboarding contribution. In case you need any help with onboarding or have any questions or feedback please reach out to our slack channel [#forum-community-operators](https://redhat.enterprise.slack.com/archives/C01UYB5E414)

Thank you,
The Collective team, Red Hat

**Comment by Allda** [comment]
@ack-bot @jwendell @mvalarh @openshift-service-mesh-bot @maistra-bot @quay-devel @mbaldessari @acornett21 @gl-distribution-oc @opentelemetrybot @konveyor-release-bot @codeflare-machine-account @shefali-malhotra @streamnativebot @vmuzikar @ndmspc @jmazzitelli @VaishnaviHire @ggeorgiev-gitlab @mkuznyetsov @raffaelespazzoli @acm-canary-bot @DanielePalaia @dt-team-kubernetes @mmontes11 @suhanime @zhiweiyin318 @che-incubator-bot @dlom @ryanemerson @jotak @gusfcarvalho @christophd @wallrj @spilchen @NissesSenap @scholzj @openshift-edge-bot @2uasimojo @pavolloffay @djzager @selvamt94 @devOpsHelm @maskarb @moolen @lbroudoux @sabre1041 @hco-bot @eguzki @iamabhishek-dubey @becash143 @spolti @tolusha @piyush-nimbalkar @abutcher

**Comment by Allda** [comment]
Here is the introduction video with a step-by-step process for onboarding the FBC. Go check that out!

https://www.youtube.com/watch?v=6yOyw0LhZ3w

And here is the list of PR with usecases mentioned in the video:

- [Existing operator conversion](https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/5035)
- [Operator removal](https://github.com/Allda/community-operators-pipeline-preprod/pull/11)
- [Operator removal from single ocp version](https://github.com/Allda/community-operators-pipeline-preprod/pull/9)
- [Channel rename](https://github.com/Allda/community-operators-pipeline-preprod/pull/8)
- [Conversion to SemVer template](https://github.com/redhat-openshift-ecosystem/certified-operators-preprod/pull/4796)
- [Simple bundle release with FBC workflow](https://github.com/redhat-openshift-ecosystem/certified-operators-preprod/pull/4795)

**Comment by 2uasimojo** [comment]
We're being told that FBC will become mandatory in February, but that it is not and will not be supported by the [upstream catalog](https://github.com/k8s-operatorhub/community-operators/).

[We](https://github.com/openshift/hive) produce bundles for both. We have tooling that Just Works today. This discussion thread just went from, "hey, check out this thing you might care about, in your own time," to not just, "you're going to have to change over," but, "you're going to have to change over and maintain two separate things now."

Not okay.

If you are on this discussion, and own a repository that publishes to both catalogs, and are similarly affected, please weigh in so we can find a better path forward.

**Comment by pursultani** [comment]
While we appreciate the potential benefits of FBC, we have some significant concerns about this change and its impact on our current workflows.

GitLab currently publishes to OperatorHub as well as RedHat Certified and Community Operators multiple times per month for each GitLab patch release. We have invested considerable time and resources in developing tooling that supports this process seamlessly. The current system works efficiently for us, allowing us to maintain a smooth and frequent release cycle across multiple catalogs.

The transition to FBC, especially with the February mandatory timeline, poses several challenges for us:

1. Dual maintenance: As mentioned in the previous comment, this change would require us to maintain two separate systems - one for the upstream catalog and another for FBC. This duplicates our efforts and increases complexity in our release process.
2. Timeline constraints: The proposed timeline for mandatory transition is not something we can easily prioritize at the moment. Our current roadmap and resource allocation don't allow for a quick pivot to a new system, especially one that requires significant changes to our established workflows.
3. Tooling investment: We've made substantial investments in our current tooling. A forced transition to FBC would potentially render much of this investment obsolete, requiring additional time and resources to develop new tools or significantly modify existing ones.
4. Operational impact: The frequent publishing we currently maintain might be disrupted during and after the transition, potentially affecting our ability to deliver timely updates to our users.

While we understand the intent behind FBC and its potential long-term benefits, the immediate impact on our operations is concerning. We would appreciate further discussion on this matter, particularly regarding:

- A more flexible timeline for transition
- Support for maintaining compatibility with both systems during a longer transition period
- Guidance on minimizing the impact on existing tooling and workflows

We believe that many other organizations publishing to multiple catalogs may face similar challenges. We hope that by raising these concerns, we can work together to find a path forward that addresses the needs of all stakeholders in the OpenShift operator ecosystem.

**Comment by ryanemerson** [comment]
> We're being told that FBC will become mandatory in February, but that it is not and will not be supported by the [upstream catalog](https://github.com/k8s-operatorhub/community-operators/).

This is my main concern. Having distinct processes adds additional overhead to our release and maintenance efforts.  Can anybody provide details of why the upstream can't/won't also move to FBC?

**Comment by Allda** [comment]
Hey folks, the FBC is not going to be mandatory. We originally thought that we would make it mandatory for new operators but decided to make it optional after getting user feedback.

You can keep producing the operator in its current form and it will be supported going forward. However, the "older" format is not very flexible and doesn't allow users to make changes in the update graph nor remove existing bundles. Since this is a community project and our team has a limited capacity any request for manually updating operators for non-fbc operators will be denied and migration to FBC would be a suggested option.

We are also considering adding an FBC for the k8s operator repository. At this point, both of these repositories rely on completely different solutions so adding support to k8s won't be an easy job. We are currently evaluating the benefits and trying to estimate the size of this effort.

Once we have more information we will get back to you.

Ales

**Comment by Allda** [comment]
Hey folks! We just introduced an FBC automated release. This feature no longer requires you to submit multiple PR, and it is very similar to the traditional non-FBC process that most of you are using and requires minimal development investment to switch.

Here is the announcement: https://github.com/redhat-openshift-ecosystem/community-operators-prod/discussions/6164

This feature should remove the main obstacle when onboarding the FBC. Give it a try and let me know your feedback.

**Comment by mariadb-stefangenov** [comment]
Hey @Allda,

From what I can tell there is a lot of information that is inaccurate now in the documentation, which should probably be updated.
Some things that are missing:
**- What is required with new FCB releases? From what I can tell a lot of data for example in the metadata/annotations.yaml is duplicated, is that used?**
**- The OCP_VERSION is no longer settable in the Makefile, how to configure that?**
**- Are we supposed to run `make fcb-onboarding` and `make catalogs` every time? From what I can gather, maybe just `make catalogs`?**
**- Where do we specify channels now? The annotations I figure are ignored, do we just specify them in the olm.channel entity?**
**- How do we do deprecations? This is something that is possible with a `deprecations.yaml` in FCB**
**- Why do we still need to push to `operators` when the fcb bundle `catalog.yaml` contains everything ( I am not 100% sure that it has scorecard test)?**
**- How do we automate this from an automated workflow perspective?**

Honestly I was initially following the official docs on FCB and managed to come up with all the needed files "after" stubmling accross the Makefile. At that point I managed to manually deploy it to OpenShift.... Now trying to figure out what is needed from me to do FCB here, has me stuck at dead-ends.

**Reply by razvan** [reply]
Hi, I found [these docs](https://redhat-openshift-ecosystem.github.io/operator-pipelines/users/fbc_workflow/) helpful when migrating the Stackable operators to FCB so you can maybe take some inspiration from them. Here is one [example](https://github.com/redhat-openshift-ecosystem/certified-operators/tree/main/operators/stackable-airflow-operator)

**Reply by mariadb-stefangenov** [reply]
He @razvan, yeah I've read these docs already. These and all the official docs on FCBs... I was wondering, how do you automate this process? What are the steps you take when you release?

**Reply by razvan** [reply]
Did you see the [auto-release page](https://redhat-openshift-ecosystem.github.io/operator-pipelines/users/fbc_autorelease/) ? I'll admit I didn't use it yet. For us, updating the version graph is now an additional step in the existing pipeline and it's not fully automated.

**Reply by mariadb-stefangenov** [reply]
Oh yeah! I am thinking to try it out.

**Reply by Allda** [reply]
@mariadb-stefangenov Hey, sorry I missed your comment. I think @razvan already covered it pretty well and shared the right links.

The workflow can be automated. We already have a couple of operator owners who do that. You don't have to run `make fcb-onboarding` on every change. This target only serves for the initial FBC onboarding. The `make catalogs` is the only command you need.

However, I strongly recommend using the auto-release feature that automates the whole process. In that case, you don't need to run any commands. Just submit a new operator bundle with a release config, and the pipeline will take care of the rest.
